File: pkg/templates/typescript/yutori-computer-use/tools/playwright-computer.ts
/**
 * Yutori n1 Playwright Computer Tool
 * 
 * Maps n1 action format to Playwright methods via CDP WebSocket connection.
 * Uses viewport-only screenshots optimized for Yutori n1's training data.
 * 
 * @see https://docs.yutori.com/reference/n1#screenshot-requirements
 */

import { chromium, type Browser, type BrowserContext, type Page } from 'playwright-core';
import type { ToolResult, N1Action } from './computer';
import { ToolError } from './computer';

const SCREENSHOT_DELAY_MS = 300;

// Key mappings from n1 output format to Playwright format
const KEY_MAP: Record<string, string> = {
  'Return': 'Enter',
  'BackSpace': 'Backspace',
  'Page_Up': 'PageUp',
  'Page_Down': 'PageDown',
};

const MODIFIER_MAP: Record<string, string> = {
  'ctrl': 'Control',
  'super': 'Meta',
  'command': 'Meta',
  'cmd': 'Meta',
};

export class PlaywrightComputerTool {
  private cdpWsUrl: string;
  private width: number;
  private height: number;
  private browser: Browser | null = null;
  private context: BrowserContext | null = null;
  private page: Page | null = null;

  constructor(cdpWsUrl: string, width = 1200, height = 800) {
    this.cdpWsUrl = cdpWsUrl;
    this.width = width;
    this.height = height;
  }

  async connect(): Promise<void> {
    if (this.browser) {
      return; // Already connected
    }

    this.browser = await chromium.connectOverCDP(this.cdpWsUrl);

    // Get existing context or create new one
    this.context = this.browser.contexts()[0];
    if (!this.context) {
      this.context = await this.browser.newContext();
    }

    // Handle new page events
    this.context.on('page', this.handleNewPage.bind(this));

    // Get existing page or create new one
    this.page = this.context.pages()[0];
    if (!this.page) {
      this.page = await this.context.newPage();
    }

    // Set viewport size to Yutori's recommended dimensions
    await this.page.setViewportSize({ width: this.width, height: this.height });
    this.page.on('close', this.handlePageClose.bind(this));
  }

  async disconnect(): Promise<void> {
    if (this.browser) {
      this.browser = null;
      this.context = null;
      this.page = null;
    }
  }

  private handleNewPage(page: Page): void {
    console.log('New page created');
    this.page = page;
    page.on('close', this.handlePageClose.bind(this));
  }

  private handlePageClose(closedPage: Page): void {
    console.log('Page closed');
    if (this.page === closedPage && this.context) {
      const pages = this.context.pages();
      if (pages.length > 0) {
        this.page = pages[pages.length - 1];
      } else {
        console.warn('Warning: All pages have been closed.');
        this.page = null;
      }
    }
  }

  private assertPage(): asserts this is { page: Page } {
    if (!this.page) {
      throw new ToolError('Page not available. Did you call connect()?');
    }
  }

  async execute(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const { action_type } = action;

    switch (action_type) {
      case 'click':
        return this.handleClick(action);
      case 'scroll':
        return this.handleScroll(action);
      case 'type':
        return this.handleType(action);
      case 'key_press':
        return this.handleKeyPress(action);
      case 'hover':
        return this.handleHover(action);
      case 'drag':
        return this.handleDrag(action);
      case 'wait':
        return this.handleWait();
      case 'refresh':
        return this.handleRefresh();
      case 'go_back':
        return this.handleGoBack();
      case 'goto_url':
        return this.handleGotoUrl(action);
      case 'read_texts_and_links':
        return this.handleReadTextsAndLinks();
      case 'stop':
        return this.handleStop(action);
      default:
        throw new ToolError(`Unknown action type: ${action_type}`);
    }
  }

  private async handleClick(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const coords = this.getCoordinates(action.center_coordinates);

    await this.page.mouse.click(coords.x, coords.y);
    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleScroll(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const coords = this.getCoordinates(action.center_coordinates);
    const direction = action.direction;
    const amount = action.amount ?? 3;

    if (!direction || !['up', 'down', 'left', 'right'].includes(direction)) {
      throw new ToolError(`Invalid scroll direction: ${direction}`);
    }

    const scrollDelta = amount * 100;

    await this.page.mouse.move(coords.x, coords.y);

    let deltaX = 0;
    let deltaY = 0;

    switch (direction) {
      case 'up':
        deltaY = -scrollDelta;
        break;
      case 'down':
        deltaY = scrollDelta;
        break;
      case 'left':
        deltaX = -scrollDelta;
        break;
      case 'right':
        deltaX = scrollDelta;
        break;
    }

    await this.page.mouse.wheel(deltaX, deltaY);
    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleType(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const text = action.text;
    if (!text) {
      throw new ToolError('text is required for type action');
    }

    if (action.clear_before_typing) {
      await this.page.keyboard.press('Control+a');
      await this.sleep(100);
      await this.page.keyboard.press('Backspace');
      await this.sleep(100);
    }

    await this.page.keyboard.type(text);

    if (action.press_enter_after) {
      await this.sleep(100);
      await this.page.keyboard.press('Enter');
    }

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleKeyPress(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const keyComb = action.key_comb;
    if (!keyComb) {
      throw new ToolError('key_comb is required for key_press action');
    }

    const mappedKey = this.mapKeyToPlaywright(keyComb);
    await this.page.keyboard.press(mappedKey);

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleHover(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const coords = this.getCoordinates(action.center_coordinates);

    await this.page.mouse.move(coords.x, coords.y);

    await this.sleep(SCREENSHOT_DELAY_MS);
    return this.screenshot();
  }

  private async handleDrag(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const startCoords = this.getCoordinates(action.start_coordinates);
    const endCoords = this.getCoordinates(action.center_coordinates);

    await this.page.mouse.move(startCoords.x, startCoords.y);
    await this.page.mouse.down();
    await this.sleep(50);
    await this.page.mouse.move(endCoords.x, endCoords.y, { steps: 12 });
    await this.page.mouse.up();

    await this.sleep(300);
    return this.screenshot();
  }

  private async handleWait(): Promise<ToolResult> {
    await this.sleep(2000);
    return this.screenshot();
  }

  private async handleRefresh(): Promise<ToolResult> {
    this.assertPage();
    await this.page.reload();
    await this.sleep(2000);
    return this.screenshot();
  }

  private async handleGoBack(): Promise<ToolResult> {
    this.assertPage();
    await this.page.goBack();
    await this.sleep(1500);
    return this.screenshot();
  }

  private async handleGotoUrl(action: N1Action): Promise<ToolResult> {
    this.assertPage();
    const url = action.url;
    if (!url) {
      throw new ToolError('url is required for goto_url action');
    }

    await this.page.goto(url);
    await this.sleep(2000);
    return this.screenshot();
  }

  private async handleReadTextsAndLinks(): Promise<ToolResult> {
    this.assertPage();
    try {
      const snapshot = await this.page.locator('body').ariaSnapshot();
      const url = this.page.url();
      const title = await this.page.title();

      const screenshotResult = await this.screenshot();

      return {
        base64Image: screenshotResult.base64Image,
        output: JSON.stringify({ url, title, snapshot }, null, 2),
      };
    } catch (error) {
      console.warn('read_texts_and_links failed:', error);
      return this.screenshot();
    }
  }

  private handleStop(action: N1Action): ToolResult {
    // Return the final answer without taking a screenshot
    return {
      output: action.answer || 'Task completed',
    };
  }

  async screenshot(): Promise<ToolResult> {
    this.assertPage();
    try {
      const buffer = await this.page.screenshot({ fullPage: false });

      return {
        base64Image: buffer.toString('base64'),
      };
    } catch (error) {
      throw new ToolError(`Failed to take screenshot: ${error}`);
    }
  }

  getCurrentUrl(): string {
    this.assertPage();
    return this.page.url();
  }

  private getCoordinates(coords?: [number, number]): { x: number; y: number } {
    if (!coords || coords.length !== 2) {
      // Default to center of viewport
      return { x: this.width / 2, y: this.height / 2 };
    }

    const [x, y] = coords;
    if (typeof x !== 'number' || typeof y !== 'number' || x < 0 || y < 0) {
      throw new ToolError(`Invalid coordinates: ${JSON.stringify(coords)}`);
    }

    return { x, y };
  }

  private mapKeyToPlaywright(key: string): string {
    // Handle modifier combinations (e.g., "ctrl+a" -> "Control+a")
    if (key.includes('+')) {
      const parts = key.split('+');
      const mappedParts = parts.map((part) => {
        const trimmed = part.trim();
        const lower = trimmed.toLowerCase();

        // Map modifier names
        if (MODIFIER_MAP[lower]) {
          return MODIFIER_MAP[lower];
        }

        // Check KEY_MAP for special keys
        return KEY_MAP[trimmed] || trimmed;
      });
      return mappedParts.join('+');
    }

    return KEY_MAP[key] || key;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: pkg/templates/python/yutori-computer-use/tools/playwright_computer.py
"""
Yutori n1 Playwright Computer Tool

Maps n1 action format to Playwright methods via CDP WebSocket connection.
Uses viewport-only screenshots optimized for Yutori n1's training data.

See: https://docs.yutori.com/reference/n1#screenshot-requirements
"""

import asyncio
import base64
import json
from typing import Optional

from playwright.async_api import async_playwright, Browser, BrowserContext, Page

from .base import ToolError, ToolResult
from .computer import N1Action

# Delay after actions before taking screenshot (in seconds for asyncio.sleep)
# Matches TypeScript SCREENSHOT_DELAY_MS = 300 (300ms = 0.3s)
SCREENSHOT_DELAY_S = 0.3

# Key mappings from n1 output format to Playwright format
KEY_MAP = {
    "Return": "Enter",
    "BackSpace": "Backspace",
    "Page_Up": "PageUp",
    "Page_Down": "PageDown",
}

MODIFIER_MAP = {
    "ctrl": "Control",
    "super": "Meta",
    "command": "Meta",
    "cmd": "Meta",
}


class PlaywrightComputerTool:
    def __init__(self, cdp_ws_url: str, width: int = 1200, height: int = 800):
        self.cdp_ws_url = cdp_ws_url
        self.width = width
        self.height = height
        self._playwright = None
        self._browser: Optional[Browser] = None
        self._context: Optional[BrowserContext] = None
        self._page: Optional[Page] = None

    async def connect(self) -> None:
        if self._browser:
            return  # Already connected

        self._playwright = await async_playwright().start()
        self._browser = await self._playwright.chromium.connect_over_cdp(self.cdp_ws_url)

        # Get existing context or create new one
        contexts = self._browser.contexts
        self._context = contexts[0] if contexts else await self._browser.new_context()

        # Handle new page events
        self._context.on("page", self._handle_new_page)

        # Get existing page or create new one
        pages = self._context.pages
        self._page = pages[0] if pages else await self._context.new_page()

        # Set viewport size to Yutori's recommended dimensions
        await self._page.set_viewport_size({"width": self.width, "height": self.height})
        self._page.on("close", self._handle_page_close)

    async def disconnect(self) -> None:
        if self._playwright:
            await self._playwright.stop()
        self._playwright = None
        self._browser = None
        self._context = None
        self._page = None

    def _handle_new_page(self, page: Page) -> None:
        print("New page created")
        self._page = page
        page.on("close", self._handle_page_close)

    def _handle_page_close(self, closed_page: Page) -> None:
        print("Page closed")
        if self._page == closed_page and self._context:
            pages = self._context.pages
            if pages:
                self._page = pages[-1]
            else:
                print("Warning: All pages have been closed.")
                self._page = None

    def _assert_page(self) -> Page:
        if not self._page:
            raise ToolError("Page not available. Did you call connect()?")
        return self._page

    async def execute(self, action: N1Action) -> ToolResult:
        action_type = action.get("action_type")

        handlers = {
            "click": self._handle_click,
            "scroll": self._handle_scroll,
            "type": self._handle_type,
            "key_press": self._handle_key_press,
            "hover": self._handle_hover,
            "drag": self._handle_drag,
            "wait": self._handle_wait,
            "refresh": self._handle_refresh,
            "go_back": self._handle_go_back,
            "goto_url": self._handle_goto_url,
            "read_texts_and_links": self._handle_read_texts_and_links,
            "stop": self._handle_stop,
        }

        handler = handlers.get(action_type)
        if not handler:
            raise ToolError(f"Unknown action type: {action_type}")

        return await handler(action)

    async def _handle_click(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        coords = self._get_coordinates(action.get("center_coordinates"))

        await page.mouse.click(coords["x"], coords["y"])
        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_scroll(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        coords = self._get_coordinates(action.get("center_coordinates"))
        direction = action.get("direction")
        amount = action.get("amount", 3)

        if direction not in ("up", "down", "left", "right"):
            raise ToolError(f"Invalid scroll direction: {direction}")

        scroll_delta = amount * 100

        await page.mouse.move(coords["x"], coords["y"])

        delta_x = 0
        delta_y = 0

        if direction == "up":
            delta_y = -scroll_delta
        elif direction == "down":
            delta_y = scroll_delta
        elif direction == "left":
            delta_x = -scroll_delta
        elif direction == "right":
            delta_x = scroll_delta

        await page.mouse.wheel(delta_x, delta_y)
        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_type(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        text = action.get("text")
        if not text:
            raise ToolError("text is required for type action")

        if action.get("clear_before_typing"):
            await page.keyboard.press("Control+a")
            await asyncio.sleep(0.1)
            await page.keyboard.press("Backspace")
            await asyncio.sleep(0.1)

        await page.keyboard.type(text)

        if action.get("press_enter_after"):
            await asyncio.sleep(0.1)
            await page.keyboard.press("Enter")

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_key_press(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        key_comb = action.get("key_comb")
        if not key_comb:
            raise ToolError("key_comb is required for key_press action")

        mapped_key = self._map_key_to_playwright(key_comb)
        await page.keyboard.press(mapped_key)

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_hover(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        coords = self._get_coordinates(action.get("center_coordinates"))

        await page.mouse.move(coords["x"], coords["y"])

        await asyncio.sleep(SCREENSHOT_DELAY_S)
        return await self.screenshot()

    async def _handle_drag(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        start_coords = self._get_coordinates(action.get("start_coordinates"))
        end_coords = self._get_coordinates(action.get("center_coordinates"))

        await page.mouse.move(start_coords["x"], start_coords["y"])
        await page.mouse.down()
        await asyncio.sleep(0.05)
        await page.mouse.move(end_coords["x"], end_coords["y"], steps=12)
        await page.mouse.up()

        await asyncio.sleep(0.3)
        return await self.screenshot()

    async def _handle_wait(self, action: N1Action) -> ToolResult:
        await asyncio.sleep(2)
        return await self.screenshot()

    async def _handle_refresh(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        await page.reload()
        await asyncio.sleep(2)
        return await self.screenshot()

    async def _handle_go_back(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        await page.go_back()
        await asyncio.sleep(1.5)
        return await self.screenshot()

    async def _handle_goto_url(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        url = action.get("url")
        if not url:
            raise ToolError("url is required for goto_url action")

        await page.goto(url)
        await asyncio.sleep(2)
        return await self.screenshot()

    async def _handle_read_texts_and_links(self, action: N1Action) -> ToolResult:
        page = self._assert_page()
        try:
            snapshot = await page.locator("body").aria_snapshot()
            url = page.url
            title = await page.title()

            screenshot_result = await self.screenshot()

            return {
                "base64_image": screenshot_result.get("base64_image", ""),
                "output": json.dumps({"url": url, "title": title, "snapshot": snapshot}, indent=2),
            }
        except Exception as e:
            print(f"read_texts_and_links failed: {e}")
            return await self.screenshot()

    async def _handle_stop(self, action: N1Action) -> ToolResult:
        return {"output": action.get("answer", "Task completed")}

    async def screenshot(self) -> ToolResult:
        page = self._assert_page()
        try:
            buffer = await page.screenshot(full_page=False)
            base64_image = base64.b64encode(buffer).decode("utf-8")
            return {"base64_image": base64_image}
        except Exception as e:
            raise ToolError(f"Failed to take screenshot: {e}")

    def get_current_url(self) -> str:
        page = self._assert_page()
        return page.url

    def _get_coordinates(
        self, coords: tuple[int, int] | list[int] | None
    ) -> dict[str, int]:
        if coords is None or len(coords) != 2:
            # Default to center of viewport
            return {"x": self.width // 2, "y": self.height // 2}

        x, y = coords
        if not isinstance(x, (int, float)) or not isinstance(y, (int, float)) or x < 0 or y < 0:
            raise ToolError(f"Invalid coordinates: {coords}")

        return {"x": int(x), "y": int(y)}

    def _map_key_to_playwright(self, key: str) -> str:
        # Handle modifier combinations (e.g., "ctrl+a" -> "Control+a")
        if "+" in key:
            parts = key.split("+")
            mapped_parts = []
            for part in parts:
                trimmed = part.strip()
                lower = trimmed.lower()

                # Map modifier names
                if lower in MODIFIER_MAP:
                    mapped_parts.append(MODIFIER_MAP[lower])
                else:
                    # Check KEY_MAP for special keys
                    mapped_parts.append(KEY_MAP.get(trimmed, trimmed))

            return "+".join(mapped_parts)

        return KEY_MAP.get(key, key)
